solutions/1.2-Introduction-to-Musig-Solutions.txt

```diff
@@ -2,75 +2,119 @@
+assert R1_digest.hex() == "aa5d4a40c1843456534d75f7246b9cbf0f825a36113102e76193f80e6c652c72"
+assert R2_digest.hex() == "6a78f2c6ca3bf1364f0575a02ef334d176df2051084c0e3062ec2af79b5b406a"
+assert R3_digest.hex() == "bc2c379a7b6ad82b40a7cf280697505ccc039370e50eb155324dcac3b5faa6a5"
 
 # Aggregate nonces
-# Method: use aggregate_schnorr_nonces()
-# Negate individual nonces if necessary
-R_agg, negated = aggregate_schnorr_nonces([R0, R1, R2])
-if negated:
-    k0.negate()
+# Tip: Add the individual nonce points together. If the aggregated nonce is not a quadratic residue
+# then negate the aggregate nonce and individual nonce scalars and nonce points.
+R_agg = sum([R1, R2, R3])
+if jacobi_symbol(R_agg.get_y(), SECP256K1_FIELD_SIZE) != 1:
+    R_agg.negate()
     k1.negate()
     k2.negate()
+    k3.negate()
 
-print("Individual nonces are {}, {}, {}.\n".format(k0.secret, k1.secret, k2.secret))
-print("Aggregate nonce point is {}\n".format(R_agg.get_bytes().hex()))
-print("R_agg was negated:", negated)
+print("Individual nonce scalars:\n\t{}, \n\t{}, \n\t{}.\n".format(k1, k2, k3))
+print("Aggregate nonce point: {}\n".format(R_agg))
+
+# Test your solution against the aggregate_schnorr_nonces() helper function.
+# aggregate_schnorr_nonces() aggregates the nonces and returns whether the individual nonces need to be negated.
+test_k1, test_k2, test_k3 = k1, k2, k3
+test_R_agg, negated = aggregate_schnorr_nonces([R1, R2, R3])
+if negated:
+    test_k1.negate()
+    test_k2.negate()
+    test_k3.negate()
+
+assert R_agg == test_R_agg
+assert k1 == test_k1
+assert k2 == test_k2
+assert k3 == test_k3
+
+print("Success!")
 ```
 
 #### 1.2.3 Compute aggregated MuSig signature
 
 ```
-msg = hashlib.sha256(b'transaction').digest()
+msg = sha256(b'transaction')
 
 # Generate partial signatures
-s0 = sign_musig(privkey0_c, k0, R_agg, pubkey_musig, msg)
-s1 = sign_musig(privkey1_c, k1, R_agg, pubkey_musig, msg)
-s2 = sign_musig(privkey2_c, k2, R_agg, pubkey_musig, msg)
+# Method: use sign_musig() with:
+#     - individual (tweaked) privkey
+#     - individual nonce scalar
+#     - aggregate nonce point
+#     - aggregate pubkey
+#     - msg
+s1 = sign_musig(privkey1_c, k1, R_agg, pubkey_agg, msg)
+s2 = sign_musig(privkey2_c, k2, R_agg, pubkey_agg, msg)
+s3 = sign_musig(privkey3_c, k3, R_agg, pubkey_agg, msg)
+
+print("Partial signatures:\n\t{}\n\t{}\n\t{}\n".format(s1, s2, s3))
 
-# Aggregate partial signatures
-sig_agg = aggregate_musig_signatures([s0, s1, s2], R_agg)
+# Aggregate signatures
+# Method: use aggregate_musig_signatures with list of individual signatures and the aggregate nonce point
+sig_agg = aggregate_musig_signatures([s1, s2, s3], R_agg)
+print("Aggregate signature:\n\t{}\n".format(sig_agg.hex()))
 
-# Verify signature
-assert pubkey_musig.verify_schnorr(sig_agg, msg)
-print("Success! Signature verifies against MuSig pubkey")
+# Verify signature against aggregate publick key
+assert pubkey_agg.verify_schnorr(sig_agg, msg)
```

Comment:
This is failing for me during testing

Reply:
verify_schnorr failing doesn’t appear to be isolated to this PR, will address separately

Reply:
Good catch. I've pushed a fix.